superbench/common/utils/topo_aware.py

```diff
@@ -0,0 +1,211 @@
+                    local_port = int(r.groups[0])
+                    connected_to_remote_host = r.groups[1]
+                    connected_to_remote_port = int(r.groups[2])
+                    topology[current_dev][local_port] = {
+                        connected_to_remote_host: connected_to_remote_port}
+
+
+    # Build a graph across physical hosts and switch nodes
+    Gnx = nx.Graph()
+    Gnx.add_nodes_from(topologyX)
+    for dev in topology:
+        numPorts = topology[dev]['number_of_ports']
+        if numPorts > 0:
+            for port in range(1, numPorts + 1):
+                if port in topology[dev].keys():
+                    remote_host = list(topology[dev][port].keys())[0]
+                    remote_port = list(topology[dev][port].values())[0]
+                    if topology[dev]['node_type'] == 'hca':
+                        if topology[dev]['hca_type'] == 'AN':
+                            continue
+                        src = topology[dev]['hostname']
+                    else:
+                        src = topology[dev]['metadata']
+
+                    if remote_host in topology:
+                        if topology[remote_host]['node_type'] == 'hca':
+                            if topology[remote_host]['hca_type'] == 'AN':
+                                continue
+                            dest = topology[remote_host]['hostname']
+                        else:
+                            dest = topology[remote_host]['metadata']
+
+                        Gnx.add_edge(src, dest)
+
+    all_paths_len_obj=nx.all_pairs_dijkstra_path_length(Gnx)
+    all_paths_len = dict(all_paths_len_obj)
+
+    # Generate VM pairs with different topology distance
+    # 1st row has all vm pairs with distance as 2
+    # 2nd row has all vm pairs with distance as 4
+    # 3rd row has all vm pairs with distance as 6
+    config = []
+    for nodeDistance in range(2, 7, 2):
+        visited = dict()
+        hostpairs = {}
+        row = []
+        for hostx in hosts:
+            if hostx in visited or hostx.find('ufm') != -1:
+                continue
+            visited[hostx] = 1
+            for hosty in hosts:
+                if hosty in visited  or hosty.find('ufm') != -1:
+                    continue
+                if all_paths_len[hostx][hosty] == nodeDistance:
+                    hostpairs[hostx] = hosty
+                    visited[hosty] = 1
+                    break
+        for host in hostpairs:
+            # ibnetdiscover output is from physical host level,
+            # there could be some physical hosts (and its attached HCAs devices) shown
+            # in ibnetdiscover output, but do not have VM running on them.
+            # Therefore, the Key (sysimgguid) does not exist in sysimgguid_to_vmhost.
+            # In this case, just skip checking this physical host pair.
+            if (not (phyhost_to_sysimgguid[host] in sysimgguid_to_vmhost)
```

Comment:
if you move this check up - this will reduce the size of the graph, and can improve the dijkstra time.

Reply:
Good point. Updated in new revision. Thanks